lib/segment/benches/io_uring/simple.rs

```diff
@@ -0,0 +1,165 @@
+use segment::common::mmap_ops::transmute_from_u8_to_slice;
+use segment::data_types::vectors::VectorElementType;
+use segment::entry::entry_point::{OperationError, OperationResult};
+use segment::types::PointOffsetType;
+
+struct BufferMeta {
+    /// Sequential index of the processing point
+    pub index: usize,
+    /// Id of the point that is currently being processed
+    pub point_id: PointOffsetType,
+}
+
+struct Buffer {
+    /// Stores the buffer for the point vectors
+    pub buffer: Vec<u8>,
+    /// Stores the point ids that are currently being processed in each buffer.
+    pub meta: Option<BufferMeta>,
+}
+
+struct BufferStore {
+    /// Stores the buffer for the point vectors
+    pub buffers: Vec<Buffer>,
+}
+
+impl BufferStore {
+    pub fn new(num_buffers: usize, buffer_raw_size: usize) -> Self {
+        Self {
+            buffers: (0..num_buffers)
+                .map(|_| Buffer {
+                    buffer: vec![0; buffer_raw_size],
+                    meta: None,
+                })
+                .collect(),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn new_empty() -> Self {
+        Self { buffers: vec![] }
+    }
+}
+
+pub struct UringReader {
+    file: File,
+    buffers: BufferStore,
+    io_uring: IoUring,
+    raw_size: usize,
+    header_size: usize,
+}
+
+impl UringReader {
+    pub fn new(
+        file: File,
+        raw_size: usize,
+        header_size: usize,
+        disk_parallelism: usize,
+    ) -> OperationResult<Self> {
+        let buffers = BufferStore::new(disk_parallelism, raw_size);
+        let io_uring = IoUring::new(disk_parallelism as _)?;
+
+        Ok(Self {
+            file,
+            buffers,
+            io_uring,
+            raw_size,
+            header_size,
+        })
+    }
+
+    fn submit_and_read(
+        &mut self,
+        mut callback: impl FnMut(usize, PointOffsetType, &[VectorElementType]),
+        unused_buffer_ids: &mut Vec<usize>,
+    ) -> OperationResult<()> {
+        let buffers_count = self.buffers.buffers.len();
+        let used_buffers_count = buffers_count - unused_buffer_ids.len();
+
+        // Wait for at least one buffer to become available
+        self.io_uring.submit_and_wait(used_buffers_count)?;
+
+        let cqe = self.io_uring.completion();
+        for entry in cqe {
+            let result = entry.result();
+            if result < 0 {
+                return Err(OperationError::service_error(format!(
+                    "io_uring operation failed with {} error",
+                    result
+                )));
+            } else if (result as usize) != self.raw_size {
+                return Err(OperationError::service_error(format!(
+                    "io_uring operation returned {} bytes instead of {}",
+                    result, self.raw_size
+                )));
+            }
+
+            let buffer_id = entry.user_data() as usize;
+            let meta = self.buffers.buffers[buffer_id].meta.take().unwrap();
```

Comment:
@generall For some reason, the benchmark fails ___here___. Could you, please, take a look? 🤔